Vulkan: Use VMA for memory allocations

K0bin opened the pull request:
When trying to build this: don't forget to update the submodules.

The error posted in https://bugs.dolphin-emu.org/issues/13064#note-30 indicates that Dolphin hits the implementation specific Vulkan memory allocation limit. (Allocation limit as in number of allocations, rather than allocated memory)

To fix this, we can use a Vulkan memory allocator.
So with this PR, Dolphin joins the long list of Vulkan applications using [AMDs VMA](https://gpuopen.com/vulkan-memory-allocator/). PCSX2 and Duckstation for example also use VMA.

In theory, this should speed up memory allocations a bit, since we don't hit the system allocator for every call but I don't think Dolphin allocates frequently enough for this to be noticable at all.
Another nice side effect is that it should automatically use dedicated allocations for resources where it's beneficial. (See [VK_KHR_dedicated_allocation](https://registry.khronos.org/vulkan/specs/1.3-extensions/man/html/VK_KHR_dedicated_allocation.html) for more info. I know that this can have a decent impact on GPU performance on Nvidia hardware. That said, desktop GPUs are ridiculously powerful for the Dolphin workload anyway, so I don't think this is noticable either.

iwubcode commented:
Good to see this being attempted again.  But this needs to be tested to make sure it doesn't run into the same issues that I saw when originally testing ( #10380 ).  We shouldn't rely on "theory".

Additionally, I think the general practice is to now use a submodule for the external code.  That'll reduce the LOC.

iwubcode commented on Source/Core/VideoBackends/Vulkan/VKTexture.cpp:
```diff
@@ -85,46 +84,28 @@ std::unique_ptr<VKTexture> VKTexture::Create(const TextureConfig& tex_config, st
-          .value_or(0)};
+  VmaAllocationCreateInfo alloc_create_info = {};
+  alloc_create_info.flags = 0;
+  alloc_create_info.usage = VMA_MEMORY_USAGE_UNKNOWN;
```
Similar to how you handle priority, if this is a compute image or a render target then you'll want to do:

```
if (tex_config.IsComputeImage() || tex_config.IsRenderTarget())
  alloc_create_info.usage = VK_IMAGE_USAGE_SAMPLED_BIT | VK_IMAGE_USAGE_COLOR_ATTACHMENT_BIT;
else
  alloc_create_info.usage = VMA_MEMORY_USAGE_UNKNOWN;
```

(not sure if `UNKNOWN` is appropriate)

K0bin replied:
You're confusing [VmaAllocationCreateInfo](https://gpuopen-librariesandsdks.github.io/VulkanMemoryAllocator/html/group__group__alloc.html#gaa5846affa1e9da3800e3e78fae2305cc) with VkImageCreateInfo. Those aren't valid values for VmaAllocationCreateInfo::usage.

Besides, we're already passing the VkImageCreateInfo to vmaCreateImage, so if VMA wants to get clever with that, it already has the necessary usages.

iwubcode replied:
I used `VMA_MEMORY_USAGE_GPU_ONLY` previously for textures.  But we might need to play with it.

iwubcode replied:
@K0bin - those are valid.  Look at [the docs](https://gpuopen-librariesandsdks.github.io/VulkanMemoryAllocator/html/usage_patterns.html).  They are showing examples for `VmaAllocationCreateInfo` which is what you use here.

iwubcode replied:
@K0bin - whoops, thanks updated!

iwubcode commented on Source/Core/VideoBackends/Vulkan/StagingBuffer.cpp:
```diff
@@ -180,59 +161,46 @@ bool StagingBuffer::AllocateBuffer(STAGING_BUFFER_TYPE type, VkDeviceSize size,
-  VkMemoryRequirements requirements;
-  vkGetBufferMemoryRequirements(g_vulkan_context->GetDevice(), *out_buffer, &requirements);
+  VmaAllocationCreateInfo alloc_create_info = {};
+  alloc_create_info.flags = 0;
```
I believe this structure is used for readback.  Docs suggest this:

```
alloc_create_info.usage = VMA_MEMORY_USAGE_AUTO;
alloc_create_info.flags = VMA_ALLOCATION_CREATE_HOST_ACCESS_RANDOM_BIT |
    VMA_ALLOCATION_CREATE_MAPPED_BIT;
```

K0bin replied:
I opted to provide VkMemoryPropertyFlags manually instead of using those flags. That way we can keep the workaround for slow incoherent memory on Mali GPUs. (see GetReadbackMemoryFlags)

iwubcode replied:
I do not know how the api decides between "flags", "required flags", and "preferred flags".  Other devs may disagree but I'd rather go with best practices than trying to cater our code to a particular vendor/driver.

That's not to say we can't support Mali but I'd rather use best practices and special case Mali.

For instance:

```
#ifdef WhateverPlatformMaliIsOn (android?)
  if (MaliWorkaround)
  {
    // Do something
  }
  else
  {
    // Do whatever we want to do
  }
#else
VmaAllocationCreateInfo alloc_create_info = {};
alloc_create_info.usage = VMA_MEMORY_USAGE_AUTO;
alloc_create_info.flags = VMA_ALLOCATION_CREATE_HOST_ACCESS_RANDOM_BIT | VMA_ALLOCATION_CREATE_MAPPED_BIT;
alloc_create_info.pool = VK_NULL_HANDLE;
alloc_create_info.pUserData = nullptr;
alloc_create_info.priority = 0.0;
#endif
```

(maybe break these into functions?)

We also avoid the branching cost on machines that don't need it.

iwubcode commented on Source/Core/VideoBackends/Vulkan/VKStreamBuffer.cpp:
```diff
@@ -54,73 +52,50 @@ bool StreamBuffer::AllocateBuffer()
-  u32 memory_type_index = g_vulkan_context->GetUploadMemoryType(memory_requirements.memoryTypeBits,
-                                                                &m_coherent_mapping);
+  VmaAllocationCreateInfo alloc_create_info = {};
+  alloc_create_info.flags = 0;
```
For my implementation I used:

```
alloc_create_info.flags = VMA_ALLOCATION_CREATE_MAPPED_BIT;
alloc_create_info.usage = VMA_MEMORY_USAGE_CPU_TO_GPU;
```

K0bin replied:
`VMA_MEMORY_USAGE_CPU_TO_GPU` is deprecated.

iwubcode replied:
Ah thanks.  I was wondering why I couldn't find anything on it.  I wonder which usage pattern is best then.

iwubcode commented on Source/Core/VideoBackends/Vulkan/VKTexture.cpp:
```diff
@@ -802,43 +781,25 @@ std::pair<VkImage, VkDeviceMemory> VKStagingTexture::CreateLinearImage(StagingTe
-          .value_or(0)};
+  VmaAllocationCreateInfo alloc_create_info = {};
+  alloc_create_info.flags = 0;
+  alloc_create_info.usage = VMA_MEMORY_USAGE_UNKNOWN;
```
Suggest we try:

```
alloc_create_info.usage = VMA_MEMORY_USAGE_AUTO;
alloc_create_info.flags = VMA_ALLOCATION_CREATE_HOST_ACCESS_SEQUENTIAL_WRITE_BIT |
    VMA_ALLOCATION_CREATE_MAPPED_BIT;
```

K0bin commented:
@iwubcode I'm open to using the `VMA_ALLOCATION_CREATE_HOST_ACCESS_SEQUENTIAL_WRITE_BIT` and `VMA_ALLOCATION_CREATE_HOST_ACCESS_RANDOM_BIT` flags but that creates a problem with the Mali slow coherent memory workaround, where we prefer a coherent memory type, even if that ends up being uncached.

Btw, I'ved tried it and on my PC and phone, performance is identical. I can't see at first glance, why your attempt was slower.
I might also get rid of the manual coherency checking, because `vmaFlushAllocation` and `vmaInvalidateAllocation` check that internally.

K0bin commented:
I guess I'll also steel the pragmas from your branch to make the compilers happy.

iwubcode commented:
@K0bin -

> I'm open to using the VMA_ALLOCATION_CREATE_HOST_ACCESS_SEQUENTIAL_WRITE_BIT and VMA_ALLOCATION_CREATE_HOST_ACCESS_RANDOM_BIT flags but that creates a problem with the Mali slow coherent memory workaround, where we prefer a coherent memory type, even if that ends up being uncached.

As mentioned in my comment.  Just personally, I don't want to cater to Mali (or any vendor / driver).  **I want to write graphics code with best practices.**  If we need workarounds for specific targets then we can do that like we did with your linear workaround (I'd personally leave workarounds out but sometimes they are inevitable).

> Btw, I'ved tried it and on my PC and phone, performance is identical. I can't see at first glance, why your attempt was slower.

Compared with master?  That's good.  I don't recall why it was slower at first but after some tweaks it became on-par.  We just didn't have much benefit to merge it then.  If we have a good reason now (I think uniformity between emulators should be a good enough reason tbh) then I'm on board.

> I guess I'll also steel the pragmas from your branch to make the compilers happy.

Please do!  Thanks again for working on this.

K0bin commented:
I switched to using the VMA flags instead of picking memory flags directly and it actually seems quite a bit faster now. Not sure about that, I haven't measured it, just a quick test. My best guess is that it uses host visible VRAM now for staging buffers and the streaming buffer, which may improve GPU performance (and thus reduce the time spent waiting for the GPU to finish because of EFB peeks).

iwubcode commented on Source/Core/VideoBackends/Vulkan/StagingBuffer.cpp:
```diff
@@ -180,59 +162,53 @@ bool StagingBuffer::AllocateBuffer(STAGING_BUFFER_TYPE type, VkDeviceSize size,
+  alloc_create_info.preferredFlags = 0;
+  alloc_create_info.requiredFlags = 0;
+
+  if (DriverDetails::HasBug(DriverDetails::BUG_SLOW_CACHED_READBACK_MEMORY)) [[unlikely]]
```
Much better.  Thank you!

iwubcode commented on Source/Core/VideoBackends/Vulkan/StagingBuffer.cpp:
```diff
@@ -180,59 +162,53 @@ bool StagingBuffer::AllocateBuffer(STAGING_BUFFER_TYPE type, VkDeviceSize size,
-  res = vkAllocateMemory(g_vulkan_context->GetDevice(), &memory_allocate_info, nullptr, out_memory);
-  if (res != VK_SUCCESS)
+  VmaAllocationCreateInfo alloc_create_info = {};
+  alloc_create_info.flags = type == STAGING_BUFFER_TYPE_UPLOAD ?
```
`VMA_ALLOCATION_CREATE_MAPPED_BIT`?

K0bin replied:
We could do that but it really makes no difference.
If we pass that flag, VMA will map the resource immediately for us and we can retrieve the pointer from the vmaAllocationInfo.
Right now we just do the mapping manually with `vmaMapAllocation`.
Both are perfectly valid approaches.

I can change it if you want me to but it really doesn't matter.

iwubcode replied:
I see what you mean.  Well, on the off chance VMA does something in the future with the info, it might be nice.  Not a blocker though.

K0bin replied:
VMA already has the info that we want it to be host visible. I'll keep it as it is.

iwubcode commented on Source/Core/VideoBackends/Vulkan/VKStreamBuffer.cpp:
```diff
@@ -54,73 +52,50 @@ bool StreamBuffer::AllocateBuffer()
-  u32 memory_type_index = g_vulkan_context->GetUploadMemoryType(memory_requirements.memoryTypeBits,
-                                                                &m_coherent_mapping);
+  VmaAllocationCreateInfo alloc_create_info = {};
+  alloc_create_info.flags = VMA_ALLOCATION_CREATE_HOST_ACCESS_SEQUENTIAL_WRITE_BIT;
```
This could also potentially use `VMA_ALLOCATION_CREATE_MAPPED_BIT`

iwubcode commented on Source/Core/VideoBackends/Vulkan/VKTexture.cpp:
```diff
@@ -802,43 +781,25 @@ std::pair<VkImage, VkDeviceMemory> VKStagingTexture::CreateLinearImage(StagingTe
-                          false)
-          .value_or(0)};
+  VmaAllocationCreateInfo alloc_create_info = {};
+  alloc_create_info.flags = 0;
```
Is there a reason you stuck with no flags over the suggestion for staging?  ( `VMA_ALLOCATION_CREATE_HOST_ACCESS_SEQUENTIAL_WRITE_BIT | VMA_ALLOCATION_CREATE_MAPPED_BIT` )

K0bin replied:
We copy the linear image to the staging buffer anyway. So we don't map it.

That's technically a unnecessary copy but there's two reasons for that:
* Linear images aren't necessarily supported for a cached memory type. They *probably* are in this case but I'd have to check this and still do the copy if they aren't.
* More important reason: I wanted to keep this stupid workaround simple.

iwubcode replied:
Sorry, missed this was the workaround.  This is fine.  Thanks for your comment.

iwubcode reviewed (changes requested):
As mentioned in a comment, the vma allocator isn't a part of the Vulkan release so it does not belong in the Vulkan external.  It needs to be in its own directory (possibly with a cmake file for completeness?).  Preferably with a submodule.

iwubcode commented:
I did a bit of testing on a variety of games.   Memory utilization is higher on this branch than it is on master.

As somewhat of an aside, there must have been a regression lately.  If I try to play the opening of Opoona my card runs out of memory.  On master it returns DEVICE OUT OF MEMORY, while on this branch textures start to glitch out.  I can get it to return DEVICE OUT OF MEMORY if I try to take a screenshot.  Not particularly, elated to this branch but it is interesting.  I will have to see when things broke.

K0bin commented:
The higher memory usage can probably be addressed by using a smaller block size.

As for the other issue, did you also use my descriptor set PR?

K0bin commented:
- Reduced the block size from the default 256MiB to 64MiB to reduce memory overhead.
- Pulled in VMA as a git submodule (someone help with the build bot pls)
- Enabled VK_EXT_memory_budget when supported and set the flag for allocations to stay within the budget
- Removed coherent bools because VMA checks that internally anyway.
- Removed manual mapping and unmapping and changed it to use the CREATE_MAPPED flag instead

iwubcode commented:
> As for the other issue, did you also use my descriptor set PR? It likely doesn't really run out of memory, just out of descriptors and or descriptor sets which that PR fixes.

Yes, I tried all your PRs, none of them fix the issue.  I watch as the VRAM usage goes higher and higher until there's no more VRAM left.  I thought it was something  related to how you changed the idle wait but that didn't resolve the issue either.

I will go ahead and try out this PR again.  Haven't looked at the code yet but the changes sound very good.

iwubcode commented:
Memory is now within an acceptable difference I think.  Using the same Last Story test, I see 2.3gb for this PR and 2.2gb for master.  I also noticed without the frame limit in place master hits a high of 221% while this PR hits around 218%.  That could be explained by slight differences between master and this PR.

Opoona still runs out of memory but overall I'm pretty happy with this implementation.  I will finish my review.

K0bin commented:
>  I also noticed without the frame limit in place master hits a high of 221% while this PR hits around 218%. That could be explained by slight differences between master and this PR.

I guess we could try setting PREFER_HOST for staging buffers too and see if that makes a difference. Beyond that, we're pretty much doing the same things as far as the CPU is concerned.

> Opoona still runs out of memory 

I can reproduce that and am working on a fix. It's unrelated to this PR though.

K0bin commented:
Fix for Opoona: #11090

iwubcode reviewed (approved):
While there might be some tweaks that can be made to the flags, overall this LGTM.

Rumi-Larry commented on Source/Core/VideoBackends/Vulkan/VulkanContext.cpp:
```diff
@@ -40,6 +40,8 @@ VulkanContext::VulkanContext(VkInstance instance, VkPhysicalDevice physical_devi
 {
+  if (m_allocator != VK_NULL_HANDLE)
+    vmaDestroyAllocator(m_allocator);
   if (m_device != VK_NULL_HANDLE)
```
You could probably fuse these branches.

K0bin replied:
I could but it doesn't really matter.

tellowkrinkle commented:
> Using PREFER_HOST for the stream buffer because otherwise VMA picks a VRAM memory type which hurts performance.

Might want to test this on ubershaders.  On Metal (which doesn't have host-visible VRAM), I saw a noticeable improvement (45 → 60 fps on [prime2tri-mapslowdown.dff.zip](https://github.com/dolphin-emu/dolphin/files/9743022/prime2tri-mapslowdown.dff.zip) on my 5600M at 4x, run with VS expansion) from allocating separate host and device buffers and issuing copy commands to upload stream buffer data (but only when running exclusive ubershaders).

K0bin commented:
> from allocating separate host and device buffers and issuing copy commands to upload stream buffer data

I think that would be something to look into after this PR is merged. Master right now doesn't use VRAM either and in my testing (without uber shaders) in Mario Galaxy, it's 25% slower.

s3731997 commented:
Testing?

K0bin commented:
@s3731997 

What do you mean? If you want to test it, go ahead.

Pokechu22 commented on Source/Core/VideoBackends/Vulkan/StagingBuffer.cpp:
```diff
@@ -180,59 +151,60 @@ bool StagingBuffer::AllocateBuffer(STAGING_BUFFER_TYPE type, VkDeviceSize size,
   {
-    LOG_VULKAN_ERROR(res, "vkCreateBuffer failed: ");
-    return false;
+    // Prefer COHERENT over CACHED
+    alloc_create_info.usage = VMA_MEMORY_USAGE_UNKNOWN;
+    alloc_create_info.requiredFlags =
+        VK_MEMORY_PROPERTY_HOST_VISIBLE_BIT | VK_MEMORY_PROPERTY_HOST_COHERENT_BIT;
+    alloc_create_info.preferredFlags = VK_MEMORY_PROPERTY_HOST_CACHED_BIT;
```
Is the comment here correct? It looks like this requires COHERENT, and wants CACHED if it's available but is OK with it not being CACHED. I would assume that in this setup it's not possible to get CACHED but not COHERENT.

K0bin replied:
When it's not possible to get both at the same time, we'd rather have COHERENT than CACHED. So it prefers COHERENT over CACHED.

Pokechu22 replied:
Maybe `// Prefer COHERENT + CACHED, but if that's not possible use just COHERENT`?

Pokechu22 commented on Source/Core/VideoBackends/Vulkan/StagingBuffer.cpp:
```diff
@@ -180,59 +151,60 @@ bool StagingBuffer::AllocateBuffer(STAGING_BUFFER_TYPE type, VkDeviceSize size,
-    LOG_VULKAN_ERROR(res, "vkBindBufferMemory failed: ");
-    vkDestroyBuffer(g_vulkan_context->GetDevice(), *out_buffer, nullptr);
-    vkFreeMemory(g_vulkan_context->GetDevice(), *out_memory, nullptr);
+    LOG_VULKAN_ERROR(res, "vkCreateBuffer failed: ");
```
Should be `LOG_VULKAN_ERROR(res, "vmaCreateBuffer failed: ");`. (All other `LOG_VULKAN_ERROR` messages you added correctly use `vma`.)

Pokechu22 commented on Source/Core/VideoBackends/Vulkan/VulkanContext.h:
```diff
@@ -26,7 +26,7 @@ class VulkanContext
   // Helper method to create a Vulkan instance.
   static VkInstance CreateVulkanInstance(WindowSystemType wstype, bool enable_debug_report,
-                                         bool enable_validation_layer);
+                                         bool enable_validation_layer, u32& vk_api_version);
```
Since this is an output parameter, it should be `u32* vk_api_version` instead.

K0bin commented:
I addressed the review comments.

Pokechu22 commented on Source/Core/VideoBackends/Vulkan/VulkanContext.cpp:
```diff
@@ -756,109 +792,6 @@ void VulkanContext::DisableDebugReports()
-  if (type_index)
-    return type_index.value();
-
-  // Try for coherent memory, with any other bits set.
-  type_index = GetMemoryType(bits, COHERENT_FLAGS, false, is_coherent);
-  if (type_index)
-  {
-    WARN_LOG_FMT(VIDEO,
-                 "Strict check for upload memory properties failed, this may affect performance");
-    return type_index.value();
-  }
-
-  // Fall back to non-coherent memory.
-  WARN_LOG_FMT(
-      VIDEO,
-      "Vulkan: Failed to find a coherent memory type for uploads, this will affect performance.");
-  type_index = GetMemoryType(bits, VK_MEMORY_PROPERTY_HOST_VISIBLE_BIT, false, is_coherent);
```
Are the situations covered by these log messages still relevant, or do we now never fall back to non-coherent memory?

K0bin replied:
I added one of the warnings back in.

Pokechu22 reviewed (approved):
Looks good to me apart from one possible naming thing.

K0bin commented on Source/Core/VideoBackends/Vulkan/StagingBuffer.h:
```diff
@@ -60,7 +58,7 @@ class StagingBuffer
   // Allocates the resources needed to create a staging buffer.
   static bool AllocateBuffer(STAGING_BUFFER_TYPE type, VkDeviceSize size, VkBufferUsageFlags usage,
-                             VkBuffer* out_buffer, VkDeviceMemory* out_memory, bool* out_coherent);
+                             VkBuffer* out_buffer, VmaAllocation* alloc, char** out_map_ptr);
```
done

Pokechu22 reviewed (approved):
Untested, but looks good to me.

JMC47 commented:
I did test this and while I didn't see any major difference, I do think this is good for other users/configurations.
